### PR TITLE
BREAKING CHANGE: Persist node's current view in the database

### PR DESCRIPTION
**Pull request by 86667**
Closes https://github.com/Zilliqa/zq2/issues/1609 and https://github.com/Zilliqa/zq2/issues/1709.

For the purpose of reducing re-sync time after consensus has not been reached for a while. 

For example, before these changes a node on `view: 30` with `finalised_view: 10` would start from `view: 10` if restarted. Now, they'll start back on `view :30`. If enough time has passed since they went down then they will jump ahead the minimum number of views.

These changes make breaking changes to the db

### CODE REVIEW

File: zilliqa/src/consensus.rs [hunk 1]
```diff
@@ -2438,6 +2492,14 @@ impl Consensus {
         self.view.get_view()
     }
 
+    /// Set view in memory and update tip_info.latest_view in storage
```
Comment (JamesHinshelwood) [inline zilliqa/src/consensus.rs:2495]: Do we need to keep view in memory at all? I'd prefer a single source-of-truth unless there's a good reason not to?

File: zilliqa/src/db.rs [hunk 2]
```diff
@@ -253,6 +253,8 @@ impl Db {
                 PRIMARY KEY (address, tx_hash));
             CREATE TABLE IF NOT EXISTS tip_info (
                 latest_finalized_view INTEGER,
+                latest_view INTEGER,
+                latest_view_timestamp BLOB,
```
Comment (JamesHinshelwood) [inline zilliqa/src/db.rs:257]: nit: I quite like the convention of suffixing timestamp DB columns with `_at` - i.e. `latest_view_updated_at`?

Comment (JamesHinshelwood) [reply zilliqa/src/db.rs:257]: SQLite has a `TIMESTAMP` data type - Should we use it?

File: zilliqa/src/db.rs [hunk 3]
```diff
@@ -253,6 +253,8 @@ impl Db {
                 PRIMARY KEY (address, tx_hash));
             CREATE TABLE IF NOT EXISTS tip_info (
                 latest_finalized_view INTEGER,
+                latest_view INTEGER,
```
Comment (JamesHinshelwood) [inline zilliqa/src/db.rs:256]: I wonder if `latest_view` might be a slightly misleading name. It looks very similar to `latest_finalized_view` which is pointing to our latest finalized block, however `latest_view` won't ever point to an actual block (by design).

I can't think of anything good though, other than just `view`?

File: zilliqa/src/consensus.rs [hunk 4]
```diff
@@ -1893,10 +1946,11 @@ impl Consensus {
                     new_high_qc_block_view + 1,
                     current_high_qc_view,
                 );
+                //TODO write high_qc and current_view in one db call
```
Comment (JamesHinshelwood) [inline zilliqa/src/consensus.rs:1949]: Is it the case that we always call `set_view` before sending a `Proposal`/`View`/`NewView` for a given view?

File: zilliqa/src/db.rs (hunk 2 above)
Comment (86667) [reply zilliqa/src/db.rs]: Updated column name to `view_updated_at`. I'm not sure why weve used BLOB rather than timestamp but it does allow for simpler conversion to/from. We could move over to TIMESTAMP but i see no advantage to doing so?

File: zilliqa/src/db.rs (hunk 3 above)
Comment (86667) [reply zilliqa/src/db.rs]: Agreed, updated to `view`

File: zilliqa/src/consensus.rs [hunk 5]
```diff
@@ -279,28 +235,45 @@ impl Consensus {
...
+                        .unwrap();
+
                     trace!(
                         "recovery: high_block view {0}, finalized_number {1} , start_view {2}",
                         high_block.view(),
                         finalized_number,
                         start_view
                     );
 
+                    // If timestamp of when current view was written exists then use it to estimate the minimum number of blocks the network has moved on since shut down
```
Comment (JamesHinshelwood) [inline zilliqa/src/consensus.rs:267]: I think when doing this catch up we should also send the `NewView`s for the missed views that we didn't send. Otherwise we still need to wait for the next view to progress before the nodes will re-sync.

We could discuss and make that change in a separate PR if you prefer.

**Review by JamesHinshelwood — APPROVED** [review]
LGTM :)

File: zilliqa/src/consensus.rs (hunk 5 above)
Comment (86667) [reply zilliqa/src/consensus.rs:267]: Will create a separate issue.

**Review by JamesHinshelwood — APPROVED** [review]

**Review by JamesHinshelwood — APPROVED** [review]

**Review by JamesHinshelwood — APPROVED** [review]
